File: backend/app/api/public.py
"""Public API routes."""
from __future__ import annotations

from fastapi import APIRouter, Depends, Request, Response
from sqlalchemy.orm import Session

from ..db import get_db
from ..schemas import PublicWeek, PublicWeekSummary, RunAllRequest, RunCustomRequest, RunResponse, RunSampleRequest
from ..services import public as service

router = APIRouter(prefix="/public", tags=["public"])


@router.get("/weeks", response_model=list[PublicWeekSummary])
def list_weeks(response: Response, db: Session = Depends(get_db)):
    response.headers["Cache-Control"] = "no-store"
    return service.list_published_weeks(db)


@router.get("/weeks/current", response_model=PublicWeek | None)
def current_week(response: Response, db: Session = Depends(get_db)):
    response.headers["Cache-Control"] = "no-store"
    return service.get_current_week(db)


@router.get("/weeks/{week_id}", response_model=PublicWeek)
def get_week(week_id: int, response: Response, db: Session = Depends(get_db)):
    response.headers["Cache-Control"] = "no-store"
    return service.get_week_public(db, week_id)


@router.post("/run/sample", response_model=RunResponse)
async def run_sample(payload: RunSampleRequest, request: Request, db: Session = Depends(get_db)):
    return await service.run_sample(request, db, payload)


@router.post("/run/custom", response_model=RunResponse)
async def run_custom(payload: RunCustomRequest, request: Request, db: Session = Depends(get_db)):
    return await service.run_custom(request, db, payload)


@router.post("/run/all", response_model=RunResponse)
async def run_all(payload: RunAllRequest, request: Request, db: Session = Depends(get_db)):
    return await service.run_all(request, db, payload)

File: frontend/src/api/public.ts
import { api } from './client'
import type { PublicWeek, PublicWeekSummary, RunResponse } from '../types'

export function fetchCurrentWeek(): Promise<PublicWeek | null> {
  return api<PublicWeek | null>('/api/public/weeks/current')
}

export function fetchPublishedWeeks(): Promise<PublicWeekSummary[]> {
  return api<PublicWeekSummary[]>('/api/public/weeks')
}

export function fetchPublicWeek(weekId: number): Promise<PublicWeek> {
  return api<PublicWeek>(`/api/public/weeks/${weekId}`)
}

export function runSample(weekId: number, problemId: number, code: string, sampleIndex: number): Promise<RunResponse> {
  return api<RunResponse>('/api/public/run/sample', {
    method: 'POST',
    body: JSON.stringify({ week_id: weekId, problem_id: problemId, code, sample_index: sampleIndex })
  })
}

export function runCustom(weekId: number, problemId: number, code: string, input: string): Promise<RunResponse> {
  return api<RunResponse>('/api/public/run/custom', {
    method: 'POST',
    body: JSON.stringify({ week_id: weekId, problem_id: problemId, code, input })
  })
}

export function runAll(weekId: number, problemId: number, code: string): Promise<RunResponse> {
  return api<RunResponse>('/api/public/run/all', {
    method: 'POST',
    body: JSON.stringify({ week_id: weekId, problem_id: problemId, code })
  })
}
